File: apps/mobile/android/app/src/main/java/dev/kneecap/app/NativeBridgePlugin.java
package dev.kneecap.app;

import android.app.ActivityManager;
import android.content.Context;
import android.os.Build;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import dev.kneecap.app.stt.WhisperTranscriber;

/**
 * kneecap M3/M10 — the native half of `NativeBridge.capabilities()` (M3) and
 * `NativeBridge.transcribe()` (M10)
 * (packages/native-bridge/src/capacitor-bridge.ts). Mirrors
 * ios/App/App/NativeBridgePlugin.swift for `getDeviceInfo` — `transcribe`
 * has NO iOS counterpart registered yet (see that Swift file's own doc
 * comment for why: adding one before whisper.xcframework is actually
 * embedded would break the M3 CI build for everyone; Android's Java
 * compiles safely without its native library present, iOS's Swift would
 * not).`pickMedia`/`generateProxy`/`exportProject` remain stubbed on the TS
 * side pending M4/M9.
 *
 * Registration: unlike iOS's protocol-based auto-discovery, Capacitor Android
 * requires an explicit `registerPlugin(NativeBridgePlugin.class)` call in
 * MainActivity's onCreate BEFORE `super.onCreate()` — see MainActivity.java.
 * `name = "NativeBridge"` must match the string
 * `registerPlugin<NativeBridgePluginSpec>("NativeBridge")` uses on the TS
 * side. Unlike iOS's manually-maintained `pluginMethods` array, Capacitor
 * Android discovers `@PluginMethod`-annotated methods via reflection — no
 * separate registration step for `transcribe` below.
 */
@CapacitorPlugin(name = "NativeBridge")
public class NativeBridgePlugin extends Plugin {

    @PluginMethod
    public void getDeviceInfo(PluginCall call) {
        Context context = getContext();
        ActivityManager activityManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        ActivityManager.MemoryInfo memoryInfo = new ActivityManager.MemoryInfo();
        long ramTierMb = 0;
        if (activityManager != null) {
            activityManager.getMemoryInfo(memoryInfo);
            ramTierMb = memoryInfo.totalMem / (1024 * 1024);
        }

        JSObject result = new JSObject();
        result.put("osVersion", Build.VERSION.RELEASE);
        result.put("deviceModel", Build.MANUFACTURER + " " + Build.MODEL);
        result.put("ramTierMb", ramTierMb);
        call.resolve(result);
    }

    /**
     * kneecap M10. Real plumbing, honestly incomplete native depth — see
     * `dev.kneecap.app.stt.WhisperTranscriber`'s class doc comment for
     * exactly what is and isn't wired yet. Runs off the main thread: even
     * once the two gaps documented there are closed, `whisper_full()` is a
     * synchronous, CPU-bound native call that must never block Capacitor's
     * (UI-thread-adjacent) plugin call dispatch — plan M10 item 7, "Async
     * job with progress. Never block the UI."
     */
    @PluginMethod
    public void transcribe(PluginCall call) {
        String audioUri = call.getString("audioUri");
        String modelSize = call.getString("modelSize", "tiny");
        String languageHint = call.getString("languageHint");
        if (audioUri == null) {
            call.reject("audioUri is required", "IO_ERROR");
            return;
        }

        new Thread(() -> {
            try {
                JSObject result = WhisperTranscriber.transcribe(getContext(), audioUri, modelSize, languageHint);
                call.resolve(result);
            } catch (WhisperTranscriber.NotYetWiredException e) {
                call.reject(e.getMessage(), "NOT_IMPLEMENTED");
            } catch (Exception e) {
                call.reject("transcribe failed: " + e.getMessage(), "IO_ERROR", e);
            }
        }).start();
    }
}

File: apps/mobile/ios/App/App/NativeBridgePlugin.swift
import Foundation
import Capacitor
import UIKit

/// kneecap M3 — the native half of `NativeBridge.capabilities()`
/// (packages/native-bridge/src/capacitor-bridge.ts). This is the ONE bridge
/// method that is genuinely wired end-to-end in M3: everything else on the
/// TS side throws NOT_IMPLEMENTED pending M4/M9/M10. This plugin's whole job
/// is to prove the JS<->native round trip actually works.
///
/// M10 STATUS: no `transcribe` method has been added to this class yet, on
/// purpose — see `WhisperTranscriber.swift`'s header comment in this same
/// directory for the real (unwired) transcription code and exactly why
/// adding a method here first would break this file's CI-verified compile.
/// The Android side of M10 differs here: Java's `native` method
/// declarations compile without a `.so` present, so
/// `NativeBridgePlugin.java` DOES have a real `transcribe` method already
/// (see that file) — Swift + a missing xcframework has no equivalent safe
/// half-step.
///
/// Registration: Capacitor discovers local (non-npm) plugins that conform to
/// `CAPBridgedPlugin` via Objective-C runtime reflection — no explicit
/// registration call needed here (unlike Android's `registerPlugin(...)` in
/// MainActivity). `jsName` below ("NativeBridge") must match the string
/// `registerPlugin<NativeBridgePluginSpec>("NativeBridge")` uses on the TS
/// side (packages/native-bridge/src/capacitor-bridge.ts).
@objc(NativeBridgePlugin)
public class NativeBridgePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeBridgePlugin"
    public let jsName = "NativeBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getDeviceInfo", returnType: CAPPluginReturnPromise)
    ]

    /// The classic `uname()` trick for a real device identifier
    /// ("iPhone15,2") instead of `UIDevice.current.model`'s generic "iPhone".
    /// Falls back to the generic model name in the Simulator, where
    /// `machine` reports the host Mac's architecture instead.
    private func deviceIdentifier() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { partial, element in
            guard let value = element.value as? Int8, value != 0 else { return partial }
            return partial + String(UnicodeScalar(UInt8(value)))
        }
        if identifier.isEmpty || identifier.hasPrefix("x86_64") || identifier.hasPrefix("arm64") {
            return UIDevice.current.model
        }
        return identifier
    }

    @objc func getDeviceInfo(_ call: CAPPluginCall) {
        let physicalMemoryBytes = ProcessInfo.processInfo.physicalMemory
        let ramTierMb = Int(physicalMemoryBytes / (1024 * 1024))
        call.resolve([
            "osVersion": UIDevice.current.systemVersion,
            "deviceModel": deviceIdentifier(),
            "ramTierMb": ramTierMb
        ])
    }
}
